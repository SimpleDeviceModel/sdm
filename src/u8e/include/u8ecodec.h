/*
 * Copyright (c) 2015-2021 by Microproject LLC
 * 
 * This file is part of the Simple Device Model (SDM) framework.
 * 
 * SDM framework is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SDM framework is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with SDM framework.  If not, see <https://www.gnu.org/licenses/>.
 *
 * U8E is a library facilitating UTF-8 use in cross-platfrom
 * applications. Assuming that an application uses UTF-8 to represent
 * all text, U8E handles interaction with the environment, ensuring
 * that standard streams, file names, environment variables and
 * command line arguments are encoded properly. U8E implements text
 * codec classes which can be also used directly if needed.
 *
 * On Windows U8E works by using the wide-character versions of
 * API functions. On other platforms it uses locale-dependent
 * multibyte encoding (as reported by nl_langinfo) to interact with
 * the environment.
 *
 * This header file defines text codec classes which are used by
 * other U8E components.
 */

#ifndef U8ECODEC_H_INCLUDED
#define U8ECODEC_H_INCLUDED

#include <string>

/*
 * Supported encodings
 */

namespace u8e {
	enum Encoding {
		LocalMB, // Locale-dependent multibyte encoding (e.g. "ANSI" on Windows)
		OEM, // OEM codepage (meaningful only under Windows)
		WCharT, // wide characters (=UTF-16 on Windows, otherwise platform-dependent)
	
	// Unicode
		UTF8,
		UTF16, // architecture-dependent byte order
		UTF16LE,
		UTF16BE,

	// ISO encodings
		ISO_8859_1,
		ISO_8859_2,
		ISO_8859_3,
		ISO_8859_4,
		ISO_8859_5,
		ISO_8859_6,
		ISO_8859_7,
		ISO_8859_8,
		ISO_8859_9,
		ISO_8859_13,
		ISO_8859_15,
	
	// Windows encodings (so-called "ANSI")
		Windows1250,
		Windows1251,
		Windows1252,
		Windows1253,
		Windows1254,
		Windows1255,
		Windows1256,
		Windows1257,
		Windows1258,
	
	// DOS encodings
		CP437,
		CP850,
		CP866,
		
	// KOI-8
		KOI8R,
		KOI8U,
		
	// East Asian
		GB2312,
		GB18030,
		BIG5,
		ShiftJIS,
		EUCJP,
		EUCKR
	};
}

/*
 * Codec class definition
 */

namespace u8e {
	class CodecImpl;
	
	class Codec {
	private:
		CodecImpl *_pimpl;
	public:
		Codec(Encoding from,Encoding to);
		Codec(const Codec &)=delete;
		Codec(Codec &&orig);
		~Codec();
		
		Codec &operator=(const Codec &)=delete;
		Codec &operator=(Codec &&other);
		
		std::string transcode(const std::string &str);
		std::size_t incomplete() const;
		void reset();
	};
}

#ifdef _WIN32

/*
 * Windows-specific class to facilitate text transcoding for wide-char API calls.
 * Neither supported nor needed under POSIX since wide-char doesn't see much use there.
 */

namespace u8e {
	class WCodec {
	private:
		CodecImpl *_pimpl;
	public:
		explicit WCodec(Encoding enc);
		WCodec(const WCodec &)=delete;
		WCodec(WCodec &&orig);
		~WCodec();
		
		WCodec &operator=(const WCodec &)=delete;
		WCodec &operator=(WCodec &&other);
		
		std::string transcode(const std::wstring &str);
		std::wstring transcode(const std::string &str);
		std::size_t incomplete() const;
		std::size_t wincomplete() const;
		void reset();
	};
}

#endif

#endif
