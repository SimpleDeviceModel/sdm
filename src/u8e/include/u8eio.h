/*
 * Copyright (c) 2015-2022 Simple Device Model contributors
 * 
 * This file is part of the Simple Device Model (SDM) framework.
 * 
 * SDM framework is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SDM framework is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with SDM framework.  If not, see <https://www.gnu.org/licenses/>.
 *
 * U8E is a library facilitating UTF-8 use in cross-platfrom
 * applications. Assuming that an application uses UTF-8 to represent
 * all text, U8E handles interaction with the environment, ensuring
 * that standard streams, file names, environment variables and
 * command line arguments are encoded properly. U8E implements text
 * codec classes which can be also used directly if needed.
 *
 * On Windows U8E works by using the wide-character versions of
 * API functions. On other platforms it uses locale-dependent
 * multibyte encoding (as reported by nl_langinfo) to interact with
 * the environment.
 *
 * This header file defines the following conversion wrappers for the
 * standard I/O streams:
 *     utf8cin()
 *     utf8cout()
 *     utf8cerr()
 *
 * U8E performs standard stream encoding conversion only if the stream
 * is detected to be bound to a console/terminal, otherwise (i.e. if the
 * stream is redirected) it leaves it as it is.
 */

#ifndef U8EIO_H_INCLUDED
#define U8EIO_H_INCLUDED

#include <iostream>

namespace u8e {
	enum StreamType {NonStandard,Terminal,Redirected};
	
	StreamType streamType(std::ios_base &s);
	
	std::istream &utf8cin();
	std::ostream &utf8cout();
	std::ostream &utf8cerr();
	
	using std::endl;
	using std::flush;
}

#endif
