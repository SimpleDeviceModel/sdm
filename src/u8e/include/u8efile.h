/*
 * Copyright (c) 2015-2022 Simple Device Model contributors
 * 
 * This file is part of the Simple Device Model (SDM) framework.
 * 
 * SDM framework is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SDM framework is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with SDM framework.  If not, see <https://www.gnu.org/licenses/>.
 *
 * U8E is a library facilitating UTF-8 use in cross-platfrom
 * applications. Assuming that an application uses UTF-8 to represent
 * all text, U8E handles interaction with the environment, ensuring
 * that standard streams, file names, environment variables and
 * command line arguments are encoded properly. U8E implements text
 * codec classes which can be also used directly if needed.
 *
 * On Windows U8E works by using the wide-character versions of
 * API functions. On other platforms it uses locale-dependent
 * multibyte encoding (as reported by nl_langinfo) to interact with
 * the environment.
 *
 * This header file defines a set of iostream-derived classes similar
 * to fstream which support filename encoding conversion:
 *
 *  - OFileStream - derived from std::ostream, similar to std::ofstream
 *  - IFileStream - derived from std::istream, similar to std::ifstream
 *  - FileStream - derived from std::iostream, similar to std::fstream
 *
 * Only file names are converted, file contents are in no way altered.
 */

#ifndef U8EFILE_H_INCLUDED
#define U8EFILE_H_INCLUDED

#ifdef _MSC_VER
	#pragma warning(disable: 4996)
#endif

#include "u8ecodec.h"

#include <iostream>
#include <cstdio>

namespace u8e {
	std::FILE *cfopen(const char *filename,const char *mode,Encoding fileNameEnc=UTF8);
	
	class FileBuf : public std::streambuf {
		std::FILE *fp;
	public:
		FileBuf();
		virtual ~FileBuf();
		
		FileBuf *open(const char *filename,std::ios_base::openmode mode,Encoding fileNameEnc);
		void close();
		bool is_open() const;
	protected:
		virtual int overflow(int c);
		virtual std::streamsize xsputn(const char *s,std::streamsize n);
		
		virtual int underflow();
		virtual int uflow();
		virtual std::streamsize xsgetn(char *s,std::streamsize count);
		
		virtual std::ostream::pos_type seekoff(std::ostream::off_type off,std::ios_base::seekdir dir,std::ios_base::openmode which=std::ios_base::in|std::ios_base::out);
		virtual std::ostream::pos_type seekpos(std::ostream::pos_type pos,std::ios_base::openmode which=std::ios_base::in|std::ios_base::out);
	};
	
	class FileBufContainer {
	protected:
		FileBuf buf;
	};
	
	class OFileStream : public FileBufContainer,public std::ostream {
	public:
		OFileStream();
		explicit OFileStream(const char *filename,std::ios_base::openmode mode=std::ios_base::out,Encoding fileNameEnc=UTF8);
		
		void open(const char *filename,std::ios_base::openmode mode=std::ios_base::out,Encoding fileNameEnc=UTF8);
		void close();
		bool is_open() const;
	};
	
	class IFileStream : public FileBufContainer,public std::istream {
	public:
		IFileStream();
		explicit IFileStream(const char *filename,std::ios_base::openmode mode=std::ios_base::in,Encoding fileNameEnc=UTF8);
		
		void open(const char *filename,std::ios_base::openmode mode=std::ios_base::in,Encoding fileNameEnc=UTF8);
		void close();
		bool is_open() const;
	};
	
	class FileStream : public FileBufContainer,public std::iostream {
	public:
		FileStream();
		explicit FileStream(const char *filename,std::ios_base::openmode mode=std::ios_base::in|std::ios_base::out,Encoding fileNameEnc=UTF8);
		
		void open(const char *filename,std::ios_base::openmode mode=std::ios_base::in|std::ios_base::out,Encoding fileNameEnc=UTF8);
		void close();
		bool is_open() const;
	};
}

#endif
