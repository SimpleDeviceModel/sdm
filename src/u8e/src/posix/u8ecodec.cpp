/*
 * Copyright (c) 2015-2021 by Microproject LLC
 * 
 * This file is part of the Simple Device Model (SDM) framework.
 * 
 * SDM framework is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SDM framework is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with SDM framework.  If not, see <https://www.gnu.org/licenses/>.
 *
 * U8E is a library facilitating UTF-8 use in cross-platfrom
 * applications. Assuming that an application uses UTF-8 to represent
 * all text, U8E handles interaction with the environment, ensuring
 * that standard streams, file names, environment variables and
 * command line arguments are encoded properly. U8E implements text
 * codec classes which can be also used directly if needed.
 *
 * On Windows U8E works by using the wide-character versions of
 * API functions. On other platforms it uses locale-dependent
 * multibyte encoding (as reported by nl_langinfo) to interact with
 * the environment.
 *
 * See the corresponding header file for details.
 */

#include "u8ecodec.h"

#include <stdexcept>
#include <vector>
#include <cstring>
#include <clocale>

#include <iconv.h>
#include <errno.h>
#include <langinfo.h>

namespace {
	bool isBigEndianPlatform() {
		static const char *s="\x01\x02";
		static const int i=0x0201;
		return (memcmp(s,&i,2)!=0);
	}
	
	class Initializer {
		static Initializer _global; 
		static char _localEncodingName[256];
		
		Initializer() {
/*
 * Note: we prefer to get local encoding name during static initialization
 * since setlocale() may not be thread-safe.
 */
			localEncoding();
		}
	public:
		static const char *localEncoding() {
			if(*_localEncodingName) return _localEncodingName; // already initialized
			
			std::string oldlocale=setlocale(LC_CTYPE,NULL);
			setlocale(LC_CTYPE,"");
			strncpy(_localEncodingName,nl_langinfo(CODESET),256);
			_localEncodingName[255]=0;
			setlocale(LC_CTYPE,oldlocale.c_str());
			
// If something is wrong, assume UTF-8 and hope for the best
			if(!*_localEncodingName) strcpy(_localEncodingName,"UTF-8");
			return _localEncodingName;
		}
	};

	Initializer Initializer::_global;
	char Initializer::_localEncodingName[256];
}

/*
 * CodecImpl definition
 */

namespace u8e {
	class CodecImpl {
		template <typename T> class ignore_pointer_constness {
			T **ptr;
		public:
			ignore_pointer_constness(T** p): ptr(p) {}
			ignore_pointer_constness(const T** p): ptr(const_cast<T**>(p)) {}
			
			operator T**() {return ptr;}
			operator const T**() {return const_cast<const T**>(ptr);}
		};
		
		iconv_t codec;
		std::vector<char> vInputBuffer;
		Encoding encFrom,encTo;

// CodecImpl is uncopyable
		CodecImpl(const CodecImpl &);
		CodecImpl &operator=(const CodecImpl &);
		
		static std::string codecName(Encoding enc);
	public:
		CodecImpl(Encoding from,Encoding to);
		~CodecImpl();
		
		std::string transcode(const std::string &str);
		std::size_t incomplete() const;
		void reset();
	};
	
	CodecImpl::CodecImpl(Encoding from,Encoding to) {
		encFrom=from;
		encTo=to;
		
		codec=iconv_open(codecName(to).c_str(),codecName(from).c_str());
		
		if(codec==reinterpret_cast<iconv_t>(-1)) {
			throw std::runtime_error("Cannot create codec");
		}
	}

	CodecImpl::~CodecImpl() {
		iconv_close(codec);
	}
	
	std::string CodecImpl::codecName(Encoding enc) {
		switch(enc) {
		case LocalMB:
		case OEM:
			return Initializer::localEncoding();
		case WCharT:
			return "WCHAR_T";
		
		case UTF8:
			return "UTF-8";
		case UTF16:
			return isBigEndianPlatform()?"UTF-16BE":"UTF-16LE";
		case UTF16LE:
			return "UTF-16LE";
		case UTF16BE:
			return "UTF-16BE";
		
		case ISO_8859_1:
			return "ISO-8859-1";
		case ISO_8859_2:
			return "ISO-8859-2";
		case ISO_8859_3:
			return "ISO-8859-3";
		case ISO_8859_4:
			return "ISO-8859-4";
		case ISO_8859_5:
			return "ISO-8859-5";
		case ISO_8859_6:
			return "ISO-8859-6";
		case ISO_8859_7:
			return "ISO-8859-7";
		case ISO_8859_8:
			return "ISO-8859-8";
		case ISO_8859_9:
			return "ISO-8859-9";
		case ISO_8859_13:
			return "ISO-8859-13";
		case ISO_8859_15:
			return "ISO-8859-15";
	
		case Windows1250:
			return "CP1250";
		case Windows1251:
			return "CP1251";
		case Windows1252:
			return "CP1252";
		case Windows1253:
			return "CP1253";
		case Windows1254:
			return "CP1254";
		case Windows1255:
			return "CP1255";
		case Windows1256:
			return "CP1256";
		case Windows1257:
			return "CP1257";
		case Windows1258:
			return "CP1258";
	
		case CP437:
			return "CP437";
		case CP850:
			return "CP850";
		case CP866:
			return "CP866";
		
		case KOI8R:
			return "KOI8-R";
		case KOI8U:
			return "KOI8-U";
		
		case GB2312:
			return "GBK";
		case GB18030:
			return "GB18030";
		case BIG5:
			return "BIG5";
		case ShiftJIS:
			return "SHIFT_JIS";
		case EUCJP:
			return "EUC-JP";
		case EUCKR:
			return "EUC-KR";
		
		default:
			throw std::runtime_error("Unsupported encoding");
		}
	}
	
	std::string CodecImpl::transcode(const std::string &str) {
		const std::size_t outBufSize=256;
		char outputBuffer[outBufSize];
		std::string res;
		
		char *inbuf;
		std::size_t inbytesleft;
		char *outbuf;
		std::size_t outbytesleft;
		
		vInputBuffer.insert(vInputBuffer.end(),str.begin(),str.end());
		
		if(vInputBuffer.empty()) return std::string();
		
		inbuf=&vInputBuffer[0];
		inbytesleft=vInputBuffer.size();
		outbuf=outputBuffer;
		outbytesleft=outBufSize;
		
		for(;;) {
		// Note: 2nd argument to iconv() can be char ** or const char **, depending on platform
			std::size_t r=iconv(codec,ignore_pointer_constness<char>(&inbuf),&inbytesleft,&outbuf,&outbytesleft);
			if(r==static_cast<std::size_t>(-1)&&errno==EILSEQ) {
		// try to resynchronize
				if(inbytesleft>0) {
					inbuf++;
					inbytesleft--;
				}
			}
			res.append(outputBuffer,outBufSize-outbytesleft);
			if(r==static_cast<std::size_t>(-1)&&errno==EINVAL) {
				vInputBuffer.erase(
					vInputBuffer.begin(),
					vInputBuffer.begin()+(vInputBuffer.size()-inbytesleft)
				);
				return res;
			}
			if(inbytesleft==0) {
				vInputBuffer.clear();
				return res;
			}
			outbuf=outputBuffer;
			outbytesleft=outBufSize;
		}
	}
	
	std::size_t CodecImpl::incomplete() const {
		return vInputBuffer.size();
	}
	
	void CodecImpl::reset() {
		iconv(codec,NULL,NULL,NULL,NULL);
		vInputBuffer.clear();
	}
}

/*
 * Codec members implementation
 */

namespace u8e {
	Codec::Codec(Encoding from,Encoding to): _pimpl(new CodecImpl(from,to)) {}
	
	Codec::Codec(Codec &&orig): _pimpl(orig._pimpl) {
		orig._pimpl=nullptr;
	}
	
	Codec::~Codec() {
		delete _pimpl;
	}
	
	Codec &Codec::operator=(Codec &&other) {
		_pimpl=other._pimpl;
		other._pimpl=nullptr;
		return *this;
	}
	
	std::string Codec::transcode(const std::string &str) {
		return _pimpl->transcode(str);
	}
	
	std::size_t Codec::incomplete() const {
		return _pimpl->incomplete();
	}
	
	void Codec::reset() {
		_pimpl->reset();
	}
}
