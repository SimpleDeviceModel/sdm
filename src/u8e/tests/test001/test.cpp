#ifdef NDEBUG
	#undef NDEBUG
#endif

#include "u8ecodec.h"
#include "u8efile.h"
#include "u8eio.h"

#include <iostream>
#include <cstdlib>
#include <cassert>

using namespace u8e;

void test1() {
	utf8cout()<<"Subtest #1"<<endl;
	Codec codec(UTF8,Windows1251);
	
	IFileStream in("test.utf8.orig",std::ios::in|std::ios::binary);
	OFileStream out("test.cp1251",std::ios::out|std::ios::binary);
	
	char buf[256];
	
	for(;;) {
		in.read(buf,1+rand()%9);
		if(!in.gcount()) break;
		out<<codec.transcode(std::string(buf,static_cast<std::size_t>(in.gcount())));
	}
	
	assert(!codec.incomplete());
}

void test2() {
	utf8cout()<<"Subtest #2"<<endl;
	Codec codec(Windows1251,UTF8);
	
	IFileStream in("test.cp1251",std::ios::in|std::ios::binary);
	OFileStream out("test.utf8",std::ios::out|std::ios::binary);
	
	char buf[256];
	
	for(;;) {
		in.read(buf,1+rand()%9);
		if(!in.gcount()) break;
		out<<codec.transcode(std::string(buf,static_cast<std::size_t>(in.gcount())));
	}
	
	assert(!codec.incomplete());
}

void test3() {
	utf8cout()<<"Subtest #3"<<endl;
// Test UTF-8 auto-sync
	Codec codec(UTF8,Windows1251);
	
	const char *in="\x9F\xD1\x80\xD0\xB8\xD0\xB2\xD0\xB5\xD1\x82";
	const char *expected="\xF0\xE8\xE2\xE5\xF2";
	
	std::string res=codec.transcode(in);
	assert(res==expected);
	
	const char *in2="foo" "\xD0\x20" "bar";
	res=codec.transcode(in2);
	assert(res.find("foo")!=std::string::npos);
	assert(res.find("bar")!=std::string::npos);
}

void test4() {
	utf8cout()<<"Subtest #4"<<endl;
	Codec codec(UTF8,Windows1251);
	assert(codec.transcode("")=="");
}

void test5() {
	utf8cout()<<"Subtest #5"<<endl;
	Codec codec(UTF8,UTF16);
	codec.transcode("\xD1\x80\xD0\xB8\xD0");
	assert(codec.incomplete()==1);
	codec.reset();

	std::string s=codec.transcode("\xf0");
	assert(codec.incomplete()==1);
	assert(s.empty());
	s=codec.transcode("\x9d");
	assert(codec.incomplete()==2);
	assert(s.empty());
	s=codec.transcode("\x92");
	assert(codec.incomplete()==3);
	assert(s.empty());
	s=codec.transcode("\xb3");
	assert(codec.incomplete()==0);
}

int main() {
	test1();
	test2();
	test3();
	test4();
	test5();
	
	return 0;
}
