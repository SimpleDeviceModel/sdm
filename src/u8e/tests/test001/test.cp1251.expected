Всеобщая декларация прав человека

Принята резолюцией 217 А (III) Генеральной Ассамблеи ООН от 10 декабря 1948 года
Преамбула

Принимая во внимание, что признание достоинства, присущего всем членам человеческой семьи, и равных и неотъемлемых прав их является основой свободы, справедливости и всеобщего мира; и

принимая во внимание, что пренебрежение и презрение к правам человека привели к варварским актам, которые возмущают совесть человечества, и что создание такого мира, в котором люди будут иметь свободу слова и убеждений и будут свободны от страха и нужды, провозглашено как высокое стремление людей; и

принимая во внимание, что необходимо, чтобы права человека охранялись властью закона в целях обеспечения того, чтобы человек не был вынужден прибегать, в качестве последнего средства, к восстанию против тирании и угнетения; и

принимая во внимание, что необходимо содействовать развитию дружественных отношений между народами; и

принимая во внимание, что народы Объединенных Наций подтвердили в Уставе свою веру в основные права человека, в достоинство и ценность человеческой личности и в равноправие мужчин и женщин и решили содействовать социальному прогрессу и улучшению условий жизни при большей свободе; и

принимая во внимание, что государства-члены обязались содействовать, в сотрудничестве с Организацией Объединенных Наций, всеобщему уважению и соблюдению прав человека и основных свобод; и

принимая во внимание, что всеобщее понимание характера этих прав и свобод имеет огромное значение для полного выполнения этого обязательства,

Генеральная Ассамблея,

провозглашает настоящую Всеобщую декларацию прав человека в качестве задачи, к выполнению которой должны стремиться все народы и государства с тем, чтобы каждый человек и каждый орган общества, постоянно имея в виду настоящую Декларацию, стремились путем просвещения и образования содействовать уважению этих прав и свобод и обеспечению, путем национальных и международных прогрессивных мероприятий, всеобщего и эффективного признания и осуществления их как среди народов государств-членов Организации, так и среди народов территорий, находящихся под их юрисдикцией.
Статья 1

Все люди рождаются свободными и равными в своем достоинстве и правах. Они наделены разумом и совестью и должны поступать в отношении друг друга в духе братства.
Статья 2

Каждый человек должен обладать всеми правами и всеми свободами, провозглашенными настоящей Декларацией, без какого бы то ни было различия, как-то в отношении расы, цвета кожи, пола, языка, религии, политических или иных убеждений, национального или социального происхождения, имущественного, сословного или иного положения.

Кроме того, не должно проводиться никакого различия на основе политического, правового или международного статуса страны или территории, к которой человек принадлежит, независимо от того, является ли эта территория независимой, подопечной, несамоуправляющейся или как-либо иначе ограниченной в своем суверенитете.
Статья 3

Каждый человек имеет право на жизнь, на свободу и на личную неприкосновенность.
Статья 4

Никто не должен содержаться в рабстве или в подневольном состоянии; рабство и работорговля запрещаются во всех их видах.
Статья 5

Никто не должен подвергаться пыткам или жестоким, бесчеловечным или унижающим его достоинство обращению и наказанию.
Статья 6

Каждый человек, где бы он ни находился, имеет право на признание его правосубъектности.
Статья 7

Все люди равны перед законом и имеют право, без всякого различия, на равную защиту закона. Все люди имеют право на равную защиту от какой бы то ни было дискриминации, нарушающей настоящую Декларацию, и от какого бы то ни было подстрекательства к такой дискриминации.
Статья 8

Каждый человек имеет право на эффективное восстановление в правах компетентными национальными судами в случаях нарушения его основных прав, предоставленных ему конституцией или законом.
Статья 9

Никто не может быть подвергнут произвольному аресту, задержанию или изгнанию.
Статья 10

Каждый человек, для определения его прав и обязанностей и для установления обоснованности предъявленного ему уголовного обвинения, имеет право, на основе полного равенства, на то, чтобы его дело было рассмотрено гласно и с соблюдением всех требований справедливости независимым и беспристрастным судом.
Статья 11

1. Каждый человек, обвиняемый в совершении преступления, имеет право считаться невиновным до тех пор, пока его виновность не будет установлена законным порядком путем гласного судебного разбирательства, при котором ему обеспечиваются все возможности для защиты.

2. Никто не может быть осужден за преступление на основании совершения какого-либо деяния или за бездействие, которые во время их совершения не составляли преступления по национальным законам или по международному праву. Не может также налагаться наказание более тяжкое, нежели то, которое могло быть применено в то время, когда преступление было совершено.
Статья 12

Никто не может подвергаться произвольному вмешательству в его личную и семейную жизнь, произвольным посягательствам на неприкосновенность его жилища, тайну его корреспонденции или на его честь и репутацию. Каждый человек имеет право на защиту закона от такого вмешательства или таких посягательств.
Статья 13

1. Каждый человек имеет право свободно передвигаться и выбирать себе местожительство в пределах каждого государства.

2. Каждый человек имеет право покидать любую страну, включая свою собственную, и возвращаться в свою страну.
Статья 14

1. Каждый человек имеет право искать убежища от преследования в других странах и пользоваться этим убежищем.

2. Это право не может быть использовано в случае преследования, в действительности основанного на совершении неполитического преступления, или деяния, противоречащего целям и принципам Организации Объединенных Наций.
Статья 15

1. Каждый человек имеет право на гражданство.

2. Никто не может быть произвольно лишен своего гражданства или права изменить свое гражданство.
Статья 16

1. Мужчины и женщины, достигшие совершеннолетия, имеют право без всяких ограничений по признаку расы, национальности или религии вступать в брак и основывать свою семью. Они пользуются одинаковыми правами в отношении вступления в брак, во время состояния в браке и во время его расторжения.

2. Брак может быть заключен только при свободном и полном согласии обеих вступающих в брак сторон.

3. Семья является естественной и основной ячейкой общества и имеет право на защиту со стороны общества и государства.
Статья 17

1. Каждый человек имеет право владеть имуществом как единолично, так и совместно с другими.

2. Никто не должен быть произвольно лишен своего имущества.
Статья 18

Каждый человек имеет право на свободу мысли, совести и религии; это право включает свободу менять свою религию или убеждения и свободу исповедовать свою религию или убеждения как единолично, так и сообща с другими, публичным или частным порядком в учении, богослужении и выполнении религиозных и ритуальных обрядов.
Статья 19

Каждый человек имеет право на свободу убеждений и на свободное выражение их; это право включает свободу беспрепятственно придерживаться своих убеждений и свободу искать, получать и распространять информацию и идеи любыми средствами и независимо от государственных границ.
Статья 20

1. Каждый человек имеет право на свободу мирных собраний и ассоциаций.

2. Никто не может быть принуждаем вступать в какую-либо ассоциацию.
Статья 21

1. Каждый человек имеет право принимать участие в управлении своей страной непосредственно или через посредство свободно избранных представителей.

2. Каждый человек имеет право равного доступа к государственной службе в своей стране.

3. Воля народа должна быть основой власти правительства; эта воля должна находить себе выражение в периодических и нефальсифицированных выборах, которые должны проводиться при всеобщем и равном избирательном праве путем тайного голосования или же посредством других равнозначных форм, обеспечивающих свободу голосования.
Статья 22

Каждый человек, как член общества, имеет право на социальное обеспечение и на осуществление необходимых для поддержания его достоинства и для свободного развития его личности прав в экономической, социальной и культурной областях через посредство национальных усилий и международного сотрудничества и в соответствии со структурой и ресурсами каждого государства.
Статья 23

1. Каждый человек имеет право на труд, на свободный выбор работы, на справедливые и благоприятные условия труда и на защиту от безработицы.

2. Каждый человек, без какой-либо дискриминации, имеет право на равную оплату за равный труд.

3. Каждый работающий имеет право на справедливое и удовлетворительное вознаграждение, обеспечивающее достойное человека существование для него самого и его семьи, и дополняемое, при необходимости, другими средствами социального обеспечения.

4. Каждый человек имеет право создавать профессиональные союзы и входить в профессиональные союзы для защиты своих интересов.
Статья 24

Каждый человек имеет право на отдых и досуг, включая право на разумное ограничение рабочего дня и на оплачиваемый периодический отпуск.
Статья 25

1. Каждый человек имеет право на такой жизненный уровень, включая пищу, одежду, жилище, медицинский уход и необходимое социальное обслуживание, который необходим для поддержания здоровья и благосостояния его самого и его семьи, и право на обеспечение на случай безработицы, болезни, инвалидности, вдовства, наступления старости или иного случая утраты средств к существованию по не зависящим от него обстоятельствам.

2. Материнство и младенчество дают право на особое попечение и помощь. Все дети, родившиеся в браке или вне брака, должны пользоваться одинаковой социальной защитой.
Статья 26

1. Каждый человек имеет право на образование. Образование должно быть бесплатным по меньшей мере в том, что касается начального и общего образования. Начальное образование должно быть обязательным. Техническое и профессиональное образование должно быть общедоступным, и высшее образование должно быть одинаково доступным для всех на основе способностей каждого.

2. Образование должно быть направлено к полному развитию человеческой личности и к увеличению уважения к правам человека и основным свободам. Образование должно содействовать взаимопониманию, терпимости и дружбе между всеми народами, расовыми и религиозными группами, и должно содействовать деятельности Организации Объединенных Наций по поддержанию мира.

3. Родители имеют право приоритета в выборе вида образования для своих малолетних детей.
Статья 27

1. Каждый человек имеет право свободно участвовать в культурной жизни общества, наслаждаться искусством, участвовать в научном прогрессе и пользоваться его благами.

2. Каждый человек имеет право на защиту его моральных и материальных интересов, являющихся результатом научных, литературных или художественных трудов, автором которых он является.
Статья 28

Каждый человек имеет право на социальный и международный порядок, при котором права и свободы, изложенные в настоящей Декларации, могут быть полностью осуществлены.
Статья 29

1. Каждый человек имеет обязанности перед обществом, в котором только и возможно свободное и полное развитие его личности.

2. При осуществлении своих прав и свобод каждый человек должен подвергаться только таким ограничениям, какие установлены законом исключительно с целью обеспечения должного признания и уважения прав и свобод других и удовлетворения справедливых требований морали, общественного порядка и общего благосостояния в демократическом обществе.

3. Осуществление этих прав и свобод ни в коем случае не должно противоречить целям и принципам Организации Объединенных Наций.
Статья 30

Ничто в настоящей Декларации не может быть истолковано, как предоставление какому-либо государству, группе лиц или отдельным лицам права заниматься какой-либо деятельностью или совершать действия, направленные к уничтожению прав и свобод, изложенных в настоящей Декларации.

Null character test:  
